apps/emqx/src/emqx_config.erl

```diff
@@ -380,6 +489,127 @@ fill_defaults_for_all_roots(SchemaMod, RawConf0) ->
+                    save_configs_namespaced_tx(
+                        Namespace, RootKeyBin, OneCheckedConf, OneRawConf, Opts
+                    )
+                end
+            end,
+            maps:keys(CheckedConf)
+        )
+    end),
+    put_namespaced(Namespace, CheckedConf),
+    ok.
+
+erase_namespaced_configs(Namespace) when is_binary(Namespace) ->
+    MS = erlang:make_tuple(
+        record_info(size, ?CONFIG_TAB),
+        '_',
+        [{#?CONFIG_TAB.root_key, {Namespace, '_'}}]
+    ),
+    _ = mria:match_delete(?CONFIG_TAB, MS),
+    lists:foreach(
+        fun(RootName) ->
+            persistent_term:erase(?PERSIS_KEY(?NS_CONF(Namespace), atom(RootName)))
+        end,
+        get_root_names()
+    ),
+    ok.
+-endif.
+
+load_namespaced_configs(SchemaMod) ->
+    %% Ensure tables are ready when loading.
+    _ = emqx_config:create_tables(),
+    AllowedNSRoots = maps:keys(namespaced_config_allowed_roots()),
+    lists:foreach(
+        fun({Namespace, RootKeyBin}) ->
+            do_load_namespaced_config(SchemaMod, AllowedNSRoots, Namespace, RootKeyBin)
+        end,
+        mnesia:dirty_all_keys(?CONFIG_TAB)
+    ).
+
+do_load_namespaced_config(SchemaMod, AllowedNSRoots, Namespace, RootKeyBin) ->
+    Key = {Namespace, RootKeyBin},
+    [#?CONFIG_TAB{raw_value = RawConf0}] = mnesia:dirty_read(?CONFIG_TAB, Key),
+    RootKeyAtom = atom(RootKeyBin),
+    RawConf = #{RootKeyBin => RawConf0},
+    case check_config_namespaced(SchemaMod, RawConf, AllowedNSRoots) of
+        {ok, #{} = CheckedRoot} ->
+            CheckedConf = maps:get(RootKeyAtom, CheckedRoot, #{}),
+            clear_invalid_namespaced_config(Namespace, RootKeyBin),
+            put_namespaced(Namespace, [RootKeyAtom], CheckedConf);
+        {error, HoconErrors} ->
+            mark_namespaced_config_invalid(Namespace, RootKeyBin, HoconErrors)
+    end.
+
+mark_namespaced_config_invalid(Namespace, RootKeyBin, HoconErrors) when is_binary(Namespace) ->
+    PrevErrors = persistent_term:get(?INVALID_NS_CONF_PT_KEY(Namespace), #{}),
+    Errors = PrevErrors#{RootKeyBin => HoconErrors},
+    persistent_term:put(?INVALID_NS_CONF_PT_KEY(Namespace), Errors).
+
+get_all_namespace_config_errors() ->
+    [{Namespace, Errors} || {?INVALID_NS_CONF_PT_KEY(Namespace), Errors} <- persistent_term:get()].
+
+get_namespace_config_errors(Namespace) when is_binary(Namespace) ->
+    persistent_term:get(?INVALID_NS_CONF_PT_KEY(Namespace), undefined).
+
+clear_invalid_namespaced_config(Namespace, RootKeyBin) when is_binary(Namespace) ->
+    case get_namespace_config_errors(Namespace) of
+        undefined ->
+            ok;
+        #{} = Errors0 ->
+            Errors = maps:remove(RootKeyBin, Errors0),
+            case map_size(Errors) == 0 of
+                true ->
+                    persistent_term:erase(?INVALID_NS_CONF_PT_KEY(Namespace)),
+                    emqx_corrupt_namespace_config_checker:check(Namespace);
```

Comment:
It's a but confusing IMO: why we run "checker" after clearing the wrong config or replacing with a valid? Shouldn't it already be ok?

Reply:
Fixed to re-check if there are still errors, clear the NS otherwise.

:checkered_flag: